File: lib/constructs/networking.ts
import { Construct } from 'constructs';
import * as cdk from 'aws-cdk-lib';
import * as ec2 from 'aws-cdk-lib/aws-ec2';
import { INetworking } from '../interfaces';

export interface NetworkingProps {
  readonly ipAddresses: ec2.IIpAddresses;
  readonly vpcName?: string;
  readonly natGateways?: number;
  readonly bastionHost?: boolean;
}

export class Networking extends Construct implements INetworking {
  readonly vpc: ec2.Vpc;
  readonly bastionHost?: ec2.BastionHostLinux;
  readonly hasPrivateSubnets: boolean;

  constructor(scope: Construct, id: string, props: NetworkingProps) {
    super(scope, id);

    this.hasPrivateSubnets = props.natGateways !== 0;
    this.vpc = this.buildVpc(props);
    if (props.bastionHost) {
      this.bastionHost = new ec2.BastionHostLinux(scope, 'Bastion', {
        vpc: this.vpc,
        machineImage: ec2.MachineImage.latestAmazonLinux2(),
        instanceType: ec2.InstanceType.of(ec2.InstanceClass.T4G, ec2.InstanceSize.NANO),
        subnetSelection: this.privateSubnets ?? this.isolatedSubnets,
      });
      cdk.Tags.of(this.bastionHost.instance).add('Resource', 'Bastion');
    }
  }

  get publicSubnets(): ec2.SubnetSelection {
    return { subnetType: ec2.SubnetType.PUBLIC };
  }

  get privateSubnets(): ec2.SubnetSelection | undefined {
    if (!this.hasPrivateSubnets) {
      return undefined;
    }
    return { subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS };
  }

  get isolatedSubnets(): ec2.SubnetSelection {
    return { subnetType: ec2.SubnetType.PRIVATE_ISOLATED };
  }

  private buildVpc(props: NetworkingProps) {
    const subnetConfiguration: ec2.SubnetConfiguration[] = [
      {
        cidrMask: 24,
        name: 'public',
        subnetType: ec2.SubnetType.PUBLIC,
      },
      {
        cidrMask: 24,
        name: 'isolated',
        subnetType: ec2.SubnetType.PRIVATE_ISOLATED,
      },
    ];
    if (this.hasPrivateSubnets) {
      subnetConfiguration.push({
        cidrMask: 24,
        name: 'private',
        subnetType: ec2.SubnetType.PRIVATE_WITH_EGRESS,
      });
    }
    return new ec2.Vpc(this, 'VPC', {
      ipAddresses: props.ipAddresses,
      vpcName: props.vpcName,
      subnetConfiguration,
      natGateways: props.natGateways,
    });
  }
}

File: condensetechcdkconstructs/Networking.go
package condensetechcdkconstructs

import (
	_jsii_ "github.com/aws/jsii-runtime-go/runtime"
	_init_ "github.com/condensetech/cdk-constructs/condensetechcdkconstructs/jsii"

	"github.com/aws/aws-cdk-go/awscdk/v2/awsec2"
	"github.com/aws/constructs-go/constructs/v10"
	"github.com/condensetech/cdk-constructs/condensetechcdkconstructs/internal"
)

// The Networking construct creates a VPC which can have public, private, and isolated subnets.
//
// It enforces to define a CIDR block for the VPC, which is a best practice.
//
// If the `natGateways` property is set to a positive integer, the VPC will be created with private subnets that have access to the internet through NAT gateways.
// If instead the `natGateways` property is set to 0, the VPC will have only public and isolated subnets. In this case, the subnets will anyway use a cidrMask of `24`, so that changing the number of NAT gateways will not require to re-provision the VPC.
//
// In addition, this construct can also take care of creating a bastion host in the VPC by using the latest Amazon Linux AMI with the smallest available instance (t4g.nano), if the `bastionHostEnabled` property is set to `true`.
// Experimental.
type Networking interface {
	constructs.Construct
	INetworking
	// Returns the bastion host instance of the VPC, if any.
	// Experimental.
	BastionHost() awsec2.BastionHostLinux
	// Returns if the VPC has private subnets (with access to internet through a NAT gateway).
	// Experimental.
	HasPrivateSubnets() *bool
	// Returns the isolated subnets of the VPC (without access to internet).
	// Experimental.
	IsolatedSubnets() *awsec2.SubnetSelection
	// The tree node.
	// Experimental.
	Node() constructs.Node
	// Returns the private subnets of the VPC (with access to internet through a NAT gateway).
	// Experimental.
	PrivateSubnets() *awsec2.SubnetSelection
	// Returns the public subnets of the VPC.
	// Experimental.
	PublicSubnets() *awsec2.SubnetSelection
	// The VPC where the networking resources are created.
	// Experimental.
	Vpc() awsec2.IVpc
	// Returns a string representation of this construct.
	// Experimental.
	ToString() *string
}

// The jsii proxy struct for Networking
type jsiiProxy_Networking struct {
	internal.Type__constructsConstruct
	jsiiProxy_INetworking
}

func (j *jsiiProxy_Networking) BastionHost() awsec2.BastionHostLinux {
	var returns awsec2.BastionHostLinux
	_jsii_.Get(
		j,
		"bastionHost",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_Networking) HasPrivateSubnets() *bool {
	var returns *bool
	_jsii_.Get(
		j,
		"hasPrivateSubnets",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_Networking) IsolatedSubnets() *awsec2.SubnetSelection {
	var returns *awsec2.SubnetSelection
	_jsii_.Get(
		j,
		"isolatedSubnets",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_Networking) Node() constructs.Node {
	var returns constructs.Node
	_jsii_.Get(
		j,
		"node",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_Networking) PrivateSubnets() *awsec2.SubnetSelection {
	var returns *awsec2.SubnetSelection
	_jsii_.Get(
		j,
		"privateSubnets",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_Networking) PublicSubnets() *awsec2.SubnetSelection {
	var returns *awsec2.SubnetSelection
	_jsii_.Get(
		j,
		"publicSubnets",
		&returns,
	)
	return returns
}

func (j *jsiiProxy_Networking) Vpc() awsec2.IVpc {
	var returns awsec2.IVpc
	_jsii_.Get(
		j,
		"vpc",
		&returns,
	)
	return returns
}


// Experimental.
func NewNetworking(scope constructs.Construct, id *string, props *NetworkingProps) Networking {
	_init_.Initialize()

	if err := validateNewNetworkingParameters(scope, id, props); err != nil {
		panic(err)
	}
	j := jsiiProxy_Networking{}

	_jsii_.Create(
		"@condensetech/cdk-constructs.Networking",
		[]interface{}{scope, id, props},
		&j,
	)

	return &j
}

// Experimental.
func NewNetworking_Override(n Networking, scope constructs.Construct, id *string, props *NetworkingProps) {
	_init_.Initialize()

	_jsii_.Create(
		"@condensetech/cdk-constructs.Networking",
		[]interface{}{scope, id, props},
		n,
	)
}

// Checks if `x` is a construct.
//
// Returns: true if `x` is an object created from a class which extends `Construct`.
// Deprecated: use `x instanceof Construct` instead.
func Networking_IsConstruct(x interface{}) *bool {
	_init_.Initialize()

	if err := validateNetworking_IsConstructParameters(x); err != nil {
		panic(err)
	}
	var returns *bool

	_jsii_.StaticInvoke(
		"@condensetech/cdk-constructs.Networking",
		"isConstruct",
		[]interface{}{x},
		&returns,
	)

	return returns
}

func (n *jsiiProxy_Networking) ToString() *string {
	var returns *string

	_jsii_.Invoke(
		n,
		"toString",
		nil, // no parameters
		&returns,
	)

	return returns
}

